### ISSUE TITLE
Handle better missing source files

### ISSUE THREAD

**Issue by csoriano89** [mesonbuild/meson#841]
I forgot to include a source file in the meson.build file in a DLang application (although the source file was there), and ninja-build said:
app_test@exe/source_app_application.d.o: In function `_D5app3app11application5Usage13onThemeChangeMFC7gobject9ParamSpec9ParamSpecC7gobject7ObjectG7ObjectGZv':

Which is quite cryptic.

I believe this is because it fails when linking, however it should detect these errors like it does when the file is actually not present.

**Comment by jpakkane** [comment]
Sadly there is no way to do this in the general case. It would require either a) a language that forces a specific source layout (i.e. all sources in a directory always go into a single target) or b) parsing all of the source code to get full dependency tree. That would mean reimplementing a full parser for all languages we wish to support.

**Comment by tp-m** [comment]
Doesn't the same happen with any other build system if you leave out a source file?

The cryptic error message comes from the compiler, no? So shouldn't the compiler demangle that and show something more useful here?

**Comment by csoriano89** [comment]
@jpakkane indeed, that's why dlang has dub. Could it be possible to use this tool inside meson to detect those?

**Comment by csoriano89** [comment]
@tp-m yes that would be an option, however languages like rust and dlang has their own build system which handles this properly, instead of relying on the linker. Which also makes them not focusing on fixing the error reporting by the linker, but rather improve those build systems. I believe meson could use those tools to alleviate this issue.

**Comment by nirbheek** [comment]
What do you suggest we can do to improve this situation?

**Comment by csoriano89** [comment]
If that comment refers to me, I made my proposal in the comment just above you, no?
However, I understand if that it's not on scope for now.

**Comment by nirbheek** [comment]
What I mean is, what specifically do you think Meson can do to make the situation better? If you can point out, for instance, what D toolchains and build systems do to alleviate these issues, perhaps we can do something.
